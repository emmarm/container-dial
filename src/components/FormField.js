import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const Field = styled('div')({
  marginBottom: 15,
  position: 'relative'
});

const InputArea = styled('div')({
  alignItems: 'center',
  display: 'grid',
  gridTemplateColumns: '120px 300px',
  margin: 10
});

const Label = styled('label')({
  color: '#888',
  fontSize: 20
});

const Input = styled('input')(
  {
    padding: 8
  },
  ({ error, theme }) => ({
    outline: error && `1px solid ${theme.danger}`
  })
);

const Err = styled('p')(
  {
    fontSize: 14,
    left: 135,
    position: 'absolute',
    top: 40
  },
  ({ theme }) => ({
    color: theme.danger
  })
);

const FormField = ({
  error,
  label,
  id,
  onBlur,
  onChange,
  placeholder,
  type,
  value
}) => (
  <Field>
    <InputArea>
      <Label htmlFor={id}>{label}</Label>
      <Input
        error={!!error}
        id={id}
        onBlur={onBlur}
        onChange={onChange}
        placeholder={placeholder}
        type={type ? type : 'text'}
        value={value}
      />
    </InputArea>
    <Err>{error}</Err>
  </Field>
);

FormField.propTypes = {
  error: PropTypes.string,
  label: PropTypes.string,
  id: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  value: PropTypes.oneOf([PropTypes.string, PropTypes.number, PropTypes.bool])
};

export default FormField;
