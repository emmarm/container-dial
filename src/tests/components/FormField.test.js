import React from 'react';
import { shallow } from 'enzyme';

import FormField from '../../components/FormField';

describe('FormField component', () => {
  const props = {
    id: 'new-form',
    label: 'New Form',
    onBlur: () => {},
    onChange: () => {},
    placeholder: 'Some text',
    value: 1
  };
  const wrapper = shallow(<FormField {...props} />);

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('does not show error text if no error', () => {
    const error = wrapper
      .find('Styled(p)')
      .dive()
      .find('p')
      .text();
    expect(error).toBe('');
  });

  it('applies id to input and label htmlFor', () => {
    const label = wrapper.find('Styled(label)').first();
    const input = wrapper.find('Styled(input)').first();
    expect(label.prop('htmlFor')).toBe(props.id);
    expect(input.prop('id')).toBe(props.id);
  });

  it('applies label text', () => {
    const labelText = wrapper
      .find('Styled(label)')
      .first()
      .dive()
      .find('label')
      .text();
    expect(labelText).toBe(props.label);
  });

  it('applies placeholder text and value to input', () => {
    const input = wrapper.find('Styled(input)').first();
    expect(input.prop('placeholder')).toBe(props.placeholder);
    expect(input.prop('value')).toBe(props.value);
  });

  it('sets input type if type prop provided', () => {
    const wrapperWithType = shallow(<FormField {...props} type="number" />);
    const input = wrapperWithType.find('Styled(input)').first();
    expect(input.prop('type')).toBe('number');
  });

  it('shows error text if error', () => {
    const wrapperWithError = shallow(<FormField {...props} error="Required" />);
    const errorText = wrapperWithError
      .find('Styled(p)')
      .dive()
      .find('p')
      .text();
    expect(errorText).toBe('Required');
  });
});
