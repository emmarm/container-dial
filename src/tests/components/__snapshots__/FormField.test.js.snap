// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField component matches snapshot 1`] = `
<Styled(div)>
  <Styled(div)>
    <Styled(label)
      htmlFor="new-form"
    >
      New Form
    </Styled(label)>
    <Styled(input)
      error={false}
      id="new-form"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Some text"
      type="text"
      value={1}
    />
  </Styled(div)>
  <Styled(p) />
</Styled(div)>
`;
